### ISSUE TITLE
Feature Request - Test case parallelization when there is no class/collection fixture

### ISSUE THREAD

**Issue by jakubsuchybio** [xunit/xunit#1986]
* Did you search the documentation @ https://xunit.github.io/ ?
Yes.
* Did you search the issues list to see if someone already reported it?
Yes #1227 is kinda mentioning this, but in different context.

I propose to add new functionality for test collection to run it's test cases in parallel when and only when the test collection doesn't have any class fixture or it is not participating in collection fixture.

Why? Well xunit by default instantiate the test class for every test case. So there is no shared state when there is no class/collection fixture. Therefore these test cases can be parallelized without breaking because of some shared state.

**There are 3 possible implementations.**
A. To have this as default behaviour. Then it would be beneficial to all, but it could be too complex to grasp for beginners. That when no class/collection fixture, test cases in the same test collection runs parallel, but when class/collection fixture is present, then they run sequential.
B. Opt-out option. So basically A but when someone needs to disable it, they can via options.
C. Opt-in option. Here the default behaviour from 2x won't change, but for more "experienced" users it could bring great benefit/speedup for tests in test collections that doesn't use class/collection fixture.

Personally A without option to disable it is not feasible at all. B would be most beneficial for everyone, but for the cost of more complexity in default parallelism behaviour. So I think C is the best implementation that should definitely be in xunit 3.0.

By my humble opinion the majority of tests are in this category (not using class/collection fixture), therefore this could be really significant performance boost.

Thanks!

**Comment by andre-ss6** [comment]
I recently started with xUnit and I'm not even aware of what collections are. I have a project with many dozens tests which share absolutely no state. My expectation was that they would all be run in parallel and I was disappointed to find out they don't.

I see that in the other issue some pointed out that there would be some roadblocks with the suggested route. What about this one though? Would it be difficult to implement it this way?

**Comment by bradwilson** [comment]
We are definitely not changing the default behavior. That would unnecessarily break everybody who is already depending on it.

I'm open to adding new non-default parallelization options in v3.

**Comment by psanchez54** [comment]
An optional configuration, per project for default behaviour and per method or class to make them "parallelizable" would be awesome, I had 25 test methods for a login page and had to split it into 25 classes just for the tests to run faster in parallel...

**Comment by alexb5dh** [comment]
How about implementing something like `CollectionPerMethod` for [CollectionBehavior](https://github.com/xunit/xunit/blob/master/src/xunit.core/CollectionBehavior.cs) and keeping default as is?

**Comment by DanielHillebrandSWMS** [comment]
I would really want to have that feature to speed up my tests without having to refactor my code (place every method in a single class)

**Comment by andrejohansson** [comment]
I have a similar need, or if someone can point to me if my approach is stupid/brittle or unmotivated:

- I have a set of integration tests to run against an api
- I spin up an in memory database and an app host using CustomWebApplicationFactory
- The db and the application factory are placed in a fixture
- I share this fixture as a collection between different test classes that run against different areas of our api

So, all integration tests share context (same api hos, same db) and execute against it. But that is exactly what I want to do in this case because it closely resembles whats running in production.

But as far as I understand, I cannot make the tests run in parallell, and this becomes painfully slow as the integration test suite grows.

Whats the status on parallelization options?

**Comment by khtan** [comment]
Most developers and test writers have purposely created tests that do not share or collide with other cases, so that when a parallel test runner comes along, they can benefit from running all their tests in parallel.
This parallel implementation of XUnit.Net requires that each of these tests be placed in its own class? That does not make sense - so I am hoping I am not getting this right and there is a way to run all tests in parallel.
The excuse that current Collection functionality will break seems backwards. Why create a parallel by default solution that requires all the writers to rewrite their code? The addition of sharing context for tests in a collection is just another reason to break parallel runs. 
If this discussion has already been going on somewhere else, please provide a link so I can get up to date.

**Comment by markm77** [comment]
Even when there is a class/collection fixture that doesn't necessarily mean parallel running of tests is a problem. In my case I set up a collection fixture to instance a webhost to handle external HTTP redirects (multiple webhosts isn't really a reasonable solution) but there would still be no problem to run the tests in parallel. In fact it is very annoying that the tests can't be run in parallel as they are individually slow due to waiting on slow third-party services. They are ```[Theory]``` tests that interact with different third parties depending on ```MemberData```.

I suggest to make parallel running configurable according to the user's needs without making too many assumptions.

**Comment by Happypig375** [comment]
>We are definitely not changing the default behavior. That would unnecessarily break everybody who is already depending on it.

I'm interested in what scenarios this will break. I'd say tests requiring parallelization between method runs to be disabled are the minority and these cases should be forced to add an assembly-level attribute.

xUnit has been providing great defaults to aid in writing good tests and I will be disappointed if xUnit 3 decides to take a step in the wrong direction.

**Comment by Happypig375** [comment]
xUnit v2 already "broke customers" by running classes in parallel by default: https://xunit.net/docs/running-tests-in-parallel

It makes sense for xUnit v3 to improve on this by running methods in parallel by default since that is what the majority of cases need.

**Comment by Piedone** [comment]
@tmort93 so you're running most of the tests within collections in parallel too? Can you share how you achieved this with xUnit extension points?

**Comment by tmort93** [comment]
@Piedone Sure. I've tweaked the code a bit on the fly before posting so I apologize if something doesn't compile. 

DISCLAIMER: In writing this code it was not my intent to bastardize xUnit in any way. If I've insulted any xUnit wizards out there, I can only say that it has worked well for our needs and I've never tried to take it farther than that.

```
// This points xUnit to your custom framework and needs to be added somewhere in each xUnit project
// First parameter is fully qualified name of your test framework class
// Second parameter is name of assembly where your test framework class resides
[assembly: TestFramework("Xunit.Custom.CustomTestFramework", "Xunit.Custom")]

namespace Xunit.Custom
{
	public class CustomTestFramework : XunitTestFramework
	{
		public CustomTestFramework(IMessageSink diagnosticMessageSink)
			: base(diagnosticMessageSink) { }

		protected override ITestFrameworkExecutor CreateExecutor(AssemblyName assemblyName)
		{
			return new CustomTestExecutor(assemblyName, SourceInformationProvider, DiagnosticMessageSink);
		}
	}
}

public class CustomTestExecutor : XunitTestFrameworkExecutor
{
	private const string DEFAULT_SYNCHRONOUS_COLLECTION_NAME = nameof(DEFAULT_SYNCHRONOUS_COLLECTION_NAME);

	private readonly Dictionary<string, TestCollection> _synchronousCollections = new Dictionary<string, TestCollection>();

	public CustomTestExecutor(AssemblyName assemblyName, ISourceInformationProvider sourceInformationProvider, IMessageSink diagnosticMessageSink)
		: base(assemblyName, sourceInformationProvider, diagnosticMessageSink)
	{ }

	protected override async void RunTestCases(IEnumerable<IXunitTestCase> testCases, IMessageSink executionMessageSink, ITestFrameworkExecutionOptions executionOptions)
	{
		SetUpTestCaseParallelization(testCases);

		using var assemblyRunner = new XunitTestAssemblyRunner(TestAssembly, testCases, DiagnosticMessageSink, executionMessageSink, executionOptions);
		await assemblyRunner.RunAsync();
	}

	/// <summary>
	/// Creates a new instance of <see cref="TestCollection"/>.
	/// A new unique ID must be assigned to each new instance or tests will not execute in parallel.
	/// </summary>
	/// <param name="oldTestCollection"></param>
	/// <returns></returns>
	private static TestCollection CreateNewTestCollection(ITestCollection oldTestCollection)
	{
		var newUniqueId = Guid.NewGuid();
		var newTestCollection = new TestCollection(oldTestCollection.TestAssembly, oldTestCollection.CollectionDefinition, $"{oldTestCollection.DisplayName} {oldTestCollection.UniqueID}")
		{
			UniqueID = newUniqueId
		};
		return newTestCollection;
	}

	/// <summary>
	/// Disables parallelization for synchronous test groups that should not run in parallel.
	/// <seealso cref="SynchronousAttribute"/>
	/// </summary>
	/// <param name="oldTestCollection"></param>
	/// <param name="oldTestMethod"></param>
	/// <returns></returns>
	private TestCollection? DisableParallelizationForSynchronousTestCollections(
		ITestCollection oldTestCollection,
		ITestMethod oldTestMethod)
	{
		TestCollection? newTestCollection = null;

		var synchronousAttribute = oldTestMethod.Method
			// This is our custom attribute we created so we could group individual test methods across classes into collections that would run the tests synchronously
			.GetCustomAttributes(typeof(SynchronousAttribute)) 
			.FirstOrDefault();

		if (synchronousAttribute == null)
			return newTestCollection;

		var synchronousCollectionName = synchronousAttribute
			.GetConstructorArguments()
			.FirstOrDefault()?
			.ToString() ?? DEFAULT_SYNCHRONOUS_COLLECTION_NAME;

		if (_synchronousCollections.ContainsKey(synchronousCollectionName))
		{
			newTestCollection = _synchronousCollections[synchronousCollectionName];
		}
		else
		{
			newTestCollection = CreateNewTestCollection(oldTestCollection);
			_synchronousCollections.Add(synchronousCollectionName, newTestCollection);
		}

		return newTestCollection;
	}

	/// <summary>
	/// Default xUnit compares <see cref="TestCollection"/> instances stored in TestCase.TestMethod.TestClass.TestCollection
	/// properties. <see cref="XunitTestAssemblyRunner.OrderTestCollections"/>
	/// By default, all test cases in a test class share the same collection instance which ensures they run synchronously.
	/// By providing a unique test collection instance to every test case in a test class you can make them all run in parallel.
	/// </summary>
	/// <param name="testCases"></param>
	private void SetUpTestCaseParallelization(IEnumerable<IXunitTestCase> testCases)
	{
		var updatedTestCases = new Dictionary<string, XunitTestCase>();

		foreach (var testCase in testCases)
		{
			var oldTestMethod = testCase.TestMethod;
			var oldTestClass = oldTestMethod.TestClass;
			var oldTestCollection = oldTestMethod.TestClass.TestCollection;

			// some tests we intend to run synchronously, lets not force everything to be parallel
			if (oldTestCollection.CollectionDefinition != null || oldTestClass.Class.GetCustomAttributes(typeof(CollectionAttribute)).Any())
			{
				continue;
			}

			var newTestCollection = DisableParallelizationForSynchronousTestCollections(oldTestCollection, oldTestMethod);

			newTestCollection ??= CreateNewTestCollection(oldTestCollection);

			var newTestClass = new TestClass(newTestCollection, oldTestClass.Class);
			var newTestMethod = new TestMethod(newTestClass, oldTestMethod.Method);

			// this switch is only required if you have custom fact or theory attributes
			// otherwise just use the code for the XunitTestCase case (you won't need the second foreach,
			// below either, just bring that code up)
			// if you do have a custom attribute then you'll need to implement the TestMethodSetter
			// property inside the custom XunitTestCase types you created, I'll include the code for
			// this property below
			switch (testCase)
			{
				case CustomFactTestCase fact:
					fact.TestMethodSetter = newTestMethod;
					break;

				case CustomTheoryTestCase theory:
					theory.TestMethodSetter = newTestMethod;
					break;

				case XunitTestCase xunitTestCase:
					var newTestCase = new XunitTestCase(
						DiagnosticMessageSink,
						TestMethodDisplay.ClassAndMethod, // Shouldn't matter because this is execution, not discovery
						TestMethodDisplayOptions.None, // Shouldn't matter because this is execution, not discovery
						newTestMethod,
						xunitTestCase.TestMethodArguments);
					updatedTestCases.Add(testCase.UniqueID, newTestCase);
					break;

				default:
					throw new ArgumentOutOfRangeException($"No case for {testCase.GetType()}");
			}
		}

		foreach (var (key, value) in updatedTestCases)
		{
			var indexOfOutdatedTestCase = testCases.FindIndex(tc => tc.UniqueID == key);
			testCases[indexOfOutdatedTestCase] = value;
		}
	}
}

/// <summary>
/// Add this to each method in a group of tests that need to be run synchronously.
/// </summary>
[AttributeUsage(AttributeTargets.Method)]
public sealed class SynchronousAttribute : Attribute
{
	/// <summary>
	/// Ensure <paramref name="name"/> is the same across all methods in a particular
	/// group of tests that need to run synchronously.
	/// If you have multiple groupings of synchronous tests then provided names must be
	/// unique across all groups. Even groups in other assemblies will
	/// potentially participate in the same synchronicity if the name is the same.
	/// This depends on whether assembly parallelization is turned on in xUnit options.
	/// </summary>
	/// <param name="name"></param>
	public SynchronousAttribute(string name)
	{
		Name = name;
	}

	public string Name { get; }
}

/// <summary>
/// <see cref="TestMethod"/> has only a protected setter which
/// I need to publicly expose. Doing this circumvents needing to
/// create a new <see cref="XunitTestCase"/> in order to set my
/// new instance of <see cref="TestCollection"/> that I have created
/// to enable parallelization.
/// <seealso cref="CustomTestExecutor.SetUpTestCaseParallelization"/>
/// </summary>
public ITestMethod TestMethodSetter
{
	set => TestMethod = value;
}
```

**Comment by Piedone** [comment]
Ah interesting, thank you very much @tmort93!

**Comment by ajeckmans** [comment]
thanks @tmort93, tests in our solution are now clocking in at around 3 minutes instead of the 9 we had before. Very nice improvement!

**Comment by EnochMtzR** [comment]
@tmort93 This was such an amazing solution that I'm amazed it hasn't been added to the current implementation of the TestFramework.

## Proof of concept

This is a small proof of @tmort93's implementation working.

*Note the finished time highlighted in both cases:*

Code:

```c#
using System.Threading;
using Xunit;

[assembly: TestFramework("UnitTests.XunitExtensions.XUnitTestFrameworkWithParallelFactRunner", "UnitTests")]

namespace UnitTests;

public class UnitTest1
{
    [Fact]
    public void Test1()
    {
        Thread.Sleep(10000);
        Assert.True(true);
    }

    [Fact]
    public void Test2()
    {
        Thread.Sleep(10000);
        Assert.True(true);
    }
}
```

### Test run without parallel running:
![image](https://user-images.githubusercontent.com/31320188/150666441-de5bb8c3-a6cd-4161-bf85-a3f3e11a2b2b.png)

### Test with parallel running
![image](https://user-images.githubusercontent.com/31320188/150666517-1559cdf3-e939-4244-8c6e-6b99ee1d96a4.png)

**Comment by Piedone** [comment]
@meziantou created a NuGet for this: https://www.meziantou.net/parallelize-test-cases-execution-in-xunit.htm

**Comment by dazinator** [comment]
I think it might be good to use the Trait system to indicate parallelisation preferences.
I.e could some standard trait be introduced as metadata to nudge runners?

if xunit uses the concept of test collections, and the [approach here](https://www.meziantou.net/parallelize-test-cases-execution-in-xunit.htm) was to put tests into their own collections to force parallelisation, then the trobule with that is that:
- some tests shouldn't be parallelised.
- its better to have something explicit - preferable on the test class or method - because sequential vs parallel can have side effects and its important to be aware of that model

The benefits of the traits system is that they can also be inspected and used in filters naturally when running tests - so for example, if you wanted to run your parallel tests at a specific stage in a pipeline that becomes easier to combine in trait filters.

**Comment by bradwilson** [comment]
Full test parallelism is now available in v3 `4.0.0-pre.151` https://xunit.net/docs/using-ci-builds

Updated parallel mode documentation is available: https://xunit.net/docs/running-tests-in-parallel#parallel-modes

**Comment by CoenraadS** [comment]
If an assembly defines [assembly: Parallelization(Mode =
ParallelMode.Collections)]

And the rest runner specifies ParallelMode='all'

Which will 'win'?

**Comment by bradwilson** [comment]
> Which will 'win'?

Command line switches take precedence over configuration files, configuration files take precedence over attributes. So in this case, you would be in `all` mode.

**Comment by Piedone** [comment]
Awesome!
